File: crates/aiwatcher-server/src/execution/query.rs
//! What every query engine's managed step shares.
//!
//! ```text
//!   execute   POST {engine}{routes}/query  {pipeline, window_*?, execution_id}
//!             └─► rows ─► object store ─► ArtifactRef ─► receipt
//!   lookup    GET  {engine}{routes}/executions/{execution_id}
//!             └─► running | done {digest, rows} | absent
//! ```
//!
//! A deployment runs one query engine — Flow PHP, DataFusion or DuckDB — and
//! all three speak one contract (AW-3). So a managed step asks each of them the
//! same two questions, and what differs between `flow.rs`, `datafusion.rs` and
//! `duckdb.rs` is which kind of step each claims and which binding it reads.
//! This module is the rest.
//!
//! The script is compiled in Rust (`aiwatcher_execution::compile`); the panel's
//! direct calls to the engine are the ad-hoc path and are labelled as such.
//!
//! The address is `AIWATCHER_QUERY_URL`, and [`executors`] wires it to the one
//! engine `AIWATCHER_QUERY_ENGINE` names: a DataFusion deployment holds no Flow
//! client and so never claims a `flow_php` attempt. A step carries a script and
//! a source, never a host — a plan that named its own endpoint would be a
//! request-forgery primitive posted by anything that can reach the API.
//!
//! [`QueryClient::lookup`] asks **two** questions. The engine is the only party
//! that knows a query is *still executing*; the object store's receipt says
//! what a finished attempt produced, because no engine keeps rows. `done` with
//! no receipt is the honest gap: the query finished, the rows never landed, and
//! running it again is the only way to get them.
//!
//! **A 404 is an answer only to the lookup.** Every path a query is sent to is
//! one every engine serves, so a 404 there is some other process holding the
//! port — and read as an answer, it decodes as a table with no rows and
//! completes the step over nothing.

use std::sync::Arc;
use std::time::Duration;

use aiwatcher_datasets::QueryEngine;
use aiwatcher_execution::plan::QueryStepSpec;
use aiwatcher_execution::{
    ActivityCommand, ActivityContext, ActivityError, ActivityExecutor, ActivityResult,
    ExecutorRegistry, FailureClass, PriorAttempt,
};
use reqwest::StatusCode;
use serde::Deserialize;
use serde_json::{Value, json};

use super::artifacts::{Artifacts, Receipt, Rows, preview};
use super::flow::FlowExecutor;
use crate::config::Config;

/// The executor for the engine this deployment runs, if it has an address.
///
/// That engine's and no other's. An empty registry is a working state and says
/// so at start-up: this process then claims no query attempt, and another one
/// that does claims it instead.
#[must_use]
pub fn executors(config: &Config, artifacts: Option<&Artifacts>) -> ExecutorRegistry {
    let registry = ExecutorRegistry::new();
    let (Some(endpoint), Some(artifacts)) = (config.query_url.clone(), artifacts) else {
        return registry;
    };
    let engine = config.query_engine.as_str();
    let built: Result<Arc<dyn ActivityExecutor>, reqwest::Error> = match config.query_engine {
        QueryEngine::Flow => FlowExecutor::new(endpoint.clone(), artifacts.clone())
            .map(|executor| Arc::new(executor) as Arc<dyn ActivityExecutor>),
        QueryEngine::DataFusion | QueryEngine::DuckDb => {
            // AW-3 phases 3 and 4. Until an engine has an executor, a
            // deployment naming it claims no query step — never a Flow one.
            tracing::error!(
                %endpoint, engine,
                "this build has no executor for the query engine; no query step will be claimed"
            );
            return registry;
        }
    };
    match built {
        Ok(executor) => {
            tracing::info!(%endpoint, engine, "the work role runs managed query steps");
            registry.with(executor)
        }
        Err(error) => {
            // Not fatal, and not silent. A client that will not build is a
            // misconfiguration, and the honest consequence is that this
            // process claims no query attempt rather than failing every one
            // it claims.
            tracing::error!(
                %endpoint, engine, %error,
                "the query client could not be built; no query step will be claimed"
            );
            registry
        }
    }
}

/// One engine's client: its address, where its routes live, and the object
/// store a step's rows are written to.
#[derive(Debug)]
pub(super) struct QueryClient {
    endpoint: String,
    /// `/query`, or `/flow` for Flow during the release that renames it — see
    /// `flow.rs` for why.
    routes: &'static str,
    client: reqwest::Client,
    artifacts: Artifacts,
}

impl QueryClient {
    /// # Errors
    ///
    /// When the HTTP client cannot be built at all.
    pub(super) fn new(
        endpoint: String,
        routes: &'static str,
        artifacts: Artifacts,
    ) -> Result<Self, reqwest::Error> {
        Ok(Self {
            endpoint: endpoint.trim_end_matches('/').to_owned(),
            routes,
            // No client-wide timeout: the deadline is the *step's*, from the
            // plan, and it is applied per request. A number here would be a
            // second timeout nobody chose.
            client: reqwest::Client::builder().build()?,
            artifacts,
        })
    }

    /// Run one query step and store what it answered.
    pub(super) async fn execute(
        &self,
        spec: &QueryStepSpec,
        command: &ActivityCommand,
        context: &ActivityContext,
    ) -> Result<ActivityResult, ActivityError> {
        let key = command.idempotency_key();

        let answer: QueryAnswer = self
            .post("/query", &request_of(spec, &key), context.timeout)
            .await?
            .decode()?;

        // What the plan asked for against what the engine could do. A step
        // whose plan pinned a span and whose engine narrowed it to a duration
        // read *correct rows for a different question*: the window drifted by
        // however long the dispatch took. The rows are still the step's answer;
        // what they may not become is the answer to the key, which claims the
        // span. Believed from the answer rather than assumed from the request,
        // because only the engine knows what its sources accept — an older
        // build that never learnt `as_of` says so by omission.
        let honoured = worth_remembering(
            spec.source.window.is_some(),
            answer.window_applied.as_deref(),
            answer.deterministic,
        );
        if !answer.deterministic {
            tracing::debug!(
                key,
                "the query named a function whose value is the moment it ran; \
                 this result will not be cached"
            );
        }
        if !honoured {
            tracing::debug!(
                key,
                applied = answer.window_applied.as_deref().unwrap_or("duration"),
                "the query engine narrowed a pinned span; this result will not be cached"
            );
        }

        let rows: Rows = answer.rows;
        if answer.truncated {
            // A cap the panel is entitled to hit and a managed run is not:
            // publishing the first thousand rows of something as a dataset
            // version is how somebody draws a conclusion from a slice they did
            // not know was a slice.
            return Err(ActivityError::user_code(format!(
                "the query engine returned its {} row cap and stopped; \
                 narrow the query or raise the engine's limit",
                rows.len()
            )));
        }

        let artifact = self.artifacts.put_rows("rows", &rows).await?;
        // Data first, then the pointer to it. `aiwatcher_jobs::ORDERING`, in
        // the fifth place it applies.
        self.artifacts
            .put_receipt(&Receipt {
                idempotency_key: key,
                artifact: artifact.clone(),
                rows: rows.len(),
                runtime_digest: answer.digest.clone().unwrap_or_default(),
                stored_at: time::OffsetDateTime::now_utc(),
            })
            .await?;

        Ok(ActivityResult {
            outputs: vec![artifact],
            result: Some(preview(&answer.columns, &rows, answer.took_ms)),
            diagnostics: None,
            awaiting: None,
            cacheable: honoured,
        })
    }

    /// Whether an earlier attempt of this step already ran, and what it left.
    pub(super) async fn lookup(
        &self,
        command: &ActivityCommand,
    ) -> Result<PriorAttempt, ActivityError> {
        let key = command.idempotency_key();
        let response = self
            .get(&format!("/executions/{key}"), Duration::from_secs(10))
            .await?;

        // An engine that does not serve this route at all — an older build —
        // is an engine that cannot be asked, which is what `Absent` means. The
        // cost is a retry that may duplicate work, which is exactly the state
        // the lookup exists to leave behind — and not a reason to fail a step.
        if response.status == StatusCode::NOT_FOUND {
            return Ok(PriorAttempt::Absent);
        }

        let seen: PriorAnswer = response.decode()?;
        match seen.state.as_str() {
            "running" => Ok(PriorAttempt::Running),
            "done" => {
                let digest = seen.digest.unwrap_or_default();
                match self.stored(&key, &digest).await? {
                    Some(result) => Ok(PriorAttempt::Done(Box::new(result))),
                    // It ran and its rows never reached the store. Running it
                    // again is the only way to get them.
                    None => Ok(PriorAttempt::Absent),
                }
            }
            _ => Ok(PriorAttempt::Absent),
        }
    }

    /// The rows a finished attempt already stored, if it stored them.
    async fn stored(
        &self,
        key: &str,
        runtime_digest: &str,
    ) -> Result<Option<ActivityResult>, ActivityError> {
        let Some(receipt) = self.artifacts.receipt(key).await? else {
            return Ok(None);
        };
        if receipt.runtime_digest != runtime_digest {
            // The engine ran this key and got something else. The stored bytes
            // are a different run's, so they are not this attempt's answer —
            // and the only way to get one is to run it.
            tracing::warn!(
                key,
                stored = %receipt.runtime_digest,
                reported = %runtime_digest,
                "the query engine describes a different run of this key than the receipt does"
            );
            return Ok(None);
        }
        if !self.artifacts.holds(&receipt.artifact).await? {
            // A receipt pointing at nothing. Impossible in the ordering this
            // code keeps — data before receipt — and worth surviving anyway:
            // a bucket somebody pruned should cost a rerun, not a failure.
            return Ok(None);
        }
        Ok(Some(ActivityResult {
            outputs: vec![receipt.artifact],
            result: Some(json!({ "rows": receipt.rows, "reused": true })),
            diagnostics: None,
            awaiting: None,
            // A takeover's answer, not a fresh run: whether it may be cached
            // was decided when the attempt this receipt describes ran, and it
            // is not decided again here.
            cacheable: false,
        }))
    }

    async fn post(
        &self,
        path: &str,
        body: &Value,
        timeout: Duration,
    ) -> Result<Answered, ActivityError> {
        self.send(
            self.client.post(self.url(path)).json(body).timeout(timeout),
            Absence::IsAFailure,
        )
        .await
    }

    async fn get(&self, path: &str, timeout: Duration) -> Result<Answered, ActivityError> {
        self.send(
            self.client.get(self.url(path)).timeout(timeout),
            Absence::IsAnAnswer,
        )
        .await
    }

    fn url(&self, path: &str) -> String {
        format!("{}{}{path}", self.endpoint, self.routes)
    }

    async fn send(
        &self,
        request: reqwest::RequestBuilder,
        absence: Absence,
    ) -> Result<Answered, ActivityError> {
        let response = request.send().await.map_err(|error| {
            if error.is_timeout() {
                // Its own class, because it is the one failure that proves
                // nothing about the engine — see the module docs.
                ActivityError::timed_out(format!("the query engine did not answer: {error}"))
            } else {
                ActivityError::transient(format!("the query engine could not be reached: {error}"))
            }
        })?;
        let status = response.status();
        let body = response.text().await.map_err(|error| {
            ActivityError::transient(format!("the query engine's answer stopped short: {error}"))
        })?;
        answered(status, body, absence)
    }
}

/// The plan step, as the engine's request.
///
/// A named function rather than a literal inside `execute`, because this is the
/// seam: it is where a [`QueryStepSpec`] becomes somebody else's wire format.
/// One for every engine, because the contract is one. Pure, and tested, for
/// the same reason the script generator is.
///
/// It sends the span the plan pinned **and** the duration that span is worth,
/// so an engine that can honour bounds does, and one that cannot narrows it to
/// what it has always taken. Which of the two happened comes back on the answer
/// rather than being assumed here — see `window_applied`.
fn request_of(spec: &QueryStepSpec, key: &str) -> Value {
    let mut body = json!({
        "pipeline": spec.script,
        "execution_id": key,
    });
    if let Some(window) = spec.source.window {
        body["window_from"] = json!(window.from);
        body["window_to"] = json!(window.to);
        body["window_seconds"] = json!(window.to.saturating_sub(window.from).max(0));
    }
    body
}

/// Whether a 404 is an answer, which it is to a lookup and to nothing else.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Absence {
    IsAnAnswer,
    IsAFailure,
}

/// One response, as an answer or as the failure class that decides a retry.
fn answered(status: StatusCode, body: String, absence: Absence) -> Result<Answered, ActivityError> {
    if status.is_success() || (status == StatusCode::NOT_FOUND && absence == Absence::IsAnAnswer) {
        return Ok(Answered { status, body });
    }
    if status == StatusCode::NOT_FOUND {
        // Nothing ran: the engine never saw the query. Transient rather than
        // user code, because the fix is a port, not the pipeline — and never an
        // answer, which would decode as a table with no rows.
        return Err(ActivityError::transient(format!(
            "the query engine's address answered 404 to a query, so something else holds its \
             port ({})",
            refusal(status, &body)
        )));
    }
    // The engine's own split, kept: a query it refused is a 4xx carrying a
    // message and a column, and a 502 is aiwatcher being unreachable from
    // *there*. Only the first is the pipeline's fault.
    let class = if status.is_client_error() {
        FailureClass::UserCode
    } else {
        FailureClass::Transient
    };
    Err(ActivityError::new(class, refusal(status, &body)))
}

/// One answer, before it is decoded.
struct Answered {
    status: StatusCode,
    body: String,
}

impl Answered {
    fn decode<T: serde::de::DeserializeOwned>(self) -> Result<T, ActivityError> {
        serde_json::from_str(&self.body).map_err(|error| {
            // The shape being wrong is not transient: the same build of the
            // engine will answer identically forever, and a retry would only
            // hide which of the two is out of date.
            ActivityError::user_code(format!(
                "the query engine answered something this build cannot read: {error}"
            ))
        })
    }
}

/// What the engine said went wrong, rather than its status alone.
fn refusal(status: StatusCode, body: &str) -> String {
    #[derive(Deserialize)]
    struct Body {
        error: Detail,
    }
    #[derive(Deserialize)]
    struct Detail {
        message: String,
        #[serde(default)]
        column: usize,
    }
    match serde_json::from_str::<Body>(body) {
        Ok(refused) if refused.error.column > 0 => format!(
            "{} (at column {})",
            refused.error.message, refused.error.column
        ),
        Ok(refused) => refused.error.message,
        Err(_) => format!("the query engine answered {status}"),
    }
}

/// `POST {routes}/query`, as much of it as a managed step reads.
///
/// Deliberately not the panel's whole schema: the fields a *table on a screen*
/// needs — the grain, the dataset name, whether cells may be shortened — are
/// presentation, and a reactor that parsed them would have to keep up with
/// them.
#[derive(Debug, Deserialize)]
struct QueryAnswer {
    #[serde(default)]
    columns: Vec<String>,
    #[serde(default)]
    rows: Rows,
    #[serde(default)]
    truncated: bool,
    #[serde(default)]
    took_ms: Option<u64>,
    /// What the engine says this result hashed to, in its own encoding.
    /// Absent from an older build, which costs the cross-check in `stored` and
    /// nothing else.
    #[serde(default)]
    digest: Option<String>,
    /// `span` when the engine read the exact bounds the plan pinned,
    /// `duration` when it narrowed them to a width applied from its own now.
    ///
    /// Declared by the engine rather than inferred here, because only it knows
    /// what its sources accept. Absent from an older build, which reads as
    /// `duration` — the behaviour that build has.
    #[serde(default)]
    window_applied: Option<String>,
    /// Whether running the query again would answer the same thing.
    ///
    /// False once it named `now()`, `uuid_v4()` or another function whose value
    /// is the moment it ran, or read a corpus on disk. Only the engine can know
    /// this — the script is text until it resolves it — which is the same
    /// reason `window_applied` is declared there rather than inferred here.
    /// Defaults to true so an older Flow build, which never reported it, keeps
    /// the behaviour it had: those builds have no function that could make it
    /// false.
    #[serde(default = "yes")]
    deterministic: bool,
}

/// Serde's default for a field an older engine does not send.
fn yes() -> bool {
    true
}

/// Whether this result may be remembered under the step's cache key.
///
/// Both halves are the *engine's* answer rather than the request's: whether a
/// key is well defined is `cache_key`'s question, and whether the run that
/// produced these rows happened under those conditions is only knowable where
/// the query resolved.
fn worth_remembering(pinned_a_span: bool, applied: Option<&str>, deterministic: bool) -> bool {
    deterministic && (!pinned_a_span || applied == Some("span"))
}

/// `GET {routes}/executions/{execution_id}`.
#[derive(Debug, Deserialize)]
struct PriorAnswer {
    state: String,
    #[serde(default)]
    digest: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use aiwatcher_execution::RuntimeKind;
    use aiwatcher_execution::plan::{FlowSourceRef, ResolvedWindow};

    #[test]
    fn a_pinned_span_is_sent_as_bounds_and_as_the_width_it_is_worth() {
        // Both, so an engine that can honour bounds does and one that cannot
        // reads the field it has always read. This is the seam: extending an
        // engine is a change there, not here.
        let spec = QueryStepSpec {
            script: "data_frame()->read(runs)".to_owned(),
            source: FlowSourceRef {
                dataset: "runs".to_owned(),
                window: Some(ResolvedWindow {
                    from: 1_700_000_000,
                    to: 1_700_003_600,
                }),
                ..FlowSourceRef::default()
            },
            blocks: Vec::new(),
        };
        let body = request_of(&spec, "exec-1/read/1");
        assert_eq!(body["window_from"], 1_700_000_000_i64);
        assert_eq!(body["window_to"], 1_700_003_600_i64);
        assert_eq!(body["window_seconds"], 3600);
        assert_eq!(body["execution_id"], "exec-1/read/1");

        // And a step with no window carries none of the three, rather than a
        // zero the engine would read as "everything".
        let unwindowed = QueryStepSpec {
            source: FlowSourceRef::default(),
            ..spec
        };
        let body = request_of(&unwindowed, "exec-1/read/1");
        assert!(body.get("window_seconds").is_none());
        assert!(body.get("window_from").is_none());
    }

    #[test]
    fn a_result_is_remembered_only_when_the_engine_says_both_halves_held() {
        // A pinned span the engine honoured, over functions that answer the
        // same thing twice. Anything less is produced, reported and forgotten.
        assert!(worth_remembering(true, Some("span"), true));
        assert!(worth_remembering(false, None, true));

        // The engine narrowed the span to a width applied from its own clock,
        // so a retry five minutes later reads different rows under one key.
        assert!(!worth_remembering(true, Some("duration"), true));

        // And the half the query decides: `now()` is honest work and a cache
        // entry that would be wrong the second time it is read.
        assert!(!worth_remembering(true, Some("span"), false));
        assert!(!worth_remembering(false, None, false));
    }

    #[test]
    fn an_engine_that_never_learnt_to_report_determinism_is_believed() {
        // Those builds have no function that could make it false, so the
        // absent field means "yes" rather than "unknown" — the same reasoning
        // that lets an older engine omit `window_applied`.
        let answer: QueryAnswer = serde_json::from_str(r#"{"rows":[]}"#).expect("an empty answer");
        assert!(answer.deterministic);
    }

    #[test]
    fn a_refused_query_reports_what_the_engine_said_and_where() {
        // The message is what a reader sees beside the step, and "the engine
        // answered 422" sends them nowhere.
        let refused = refusal(
            StatusCode::UNPROCESSABLE_ENTITY,
            r#"{"error":{"message":"Unknown function 'frobnicate'.","column":42}}"#,
        );
        assert_eq!(refused, "Unknown function 'frobnicate'. (at column 42)");

        // And a body that is not the engine's shape still names the status
        // rather than saying nothing.
        assert!(
            refusal(StatusCode::BAD_GATEWAY, "<html>").contains("502"),
            "a proxy's error page is still an answer worth reporting"
        );
    }

    #[test]
    fn a_404_is_an_answer_to_a_lookup_and_never_a_table_with_no_rows() {
        // The lookup: an older build without the route is an engine that
        // cannot be asked, which is `Absent`.
        assert!(answered(StatusCode::NOT_FOUND, "{}".to_owned(), Absence::IsAnAnswer).is_ok());

        // The query: measured once against a Go service on the Flow port
        // answering `404 page not found` to everything. Decoded, that body is
        // a `QueryAnswer` with no rows, and the step completed over nothing.
        let refused = answered(
            StatusCode::NOT_FOUND,
            "404 page not found".to_owned(),
            Absence::IsAFailure,
        )
        .err()
        .expect("a 404 to a query is not an answer");
        assert!(
            matches!(refused.class, FailureClass::Transient),
            "{refused:?}"
        );
        assert!(
            refused.message.contains("holds its port"),
            "{}",
            refused.message
        );
    }

    #[test]
    fn a_deployment_holds_the_executor_for_its_engine_and_no_other() {
        let artifacts = Artifacts::new(Arc::new(
            aiwatcher_prompts::adapters::memory::MemoryObjectStore::new(),
        ));
        let flow = Config {
            query_url: Some("http://127.0.0.1:8081".to_owned()),
            ..Config::default()
        };
        assert_eq!(
            executors(&flow, Some(&artifacts)).runtimes(),
            [RuntimeKind::FlowPhp]
        );

        // A DataFusion deployment never claims a `flow_php` attempt, whatever
        // address it was given.
        let datafusion = Config {
            query_engine: QueryEngine::DataFusion,
            ..flow
        };
        assert!(
            !executors(&datafusion, Some(&artifacts))
                .runtimes()
                .contains(&RuntimeKind::FlowPhp)
        );

        // And no address is no executor: absence is a working state.
        assert!(executors(&Config::default(), Some(&artifacts)).is_empty());
    }
}

File: apps/panel/src/shared/lib/query.ts
import { useQuery, type UseQueryResult } from '@tanstack/react-query';
import { z } from 'zod';

/**
 * The client for the query engine.
 *
 * A deployment runs one — Flow PHP, DataFusion or DuckDB — chosen by
 * `AIWATCHER_QUERY_ENGINE`, and every one of them serves the same contract under
 * `/query` (AW-3). So this is one client; what differs between engines is the
 * language a query is written in, which `/query/healthz` names and
 * `useQueryEngine()` reads.
 *
 * Hand-written, unlike everything under `api/generated`: the engine is not in
 * the Rust OpenAPI document and deliberately so — the panel talks to it
 * directly, and aiwatcher's binary does not know it exists (ADR_0008). That
 * makes runtime validation the right call here for the same reason it is for
 * the SSE frames: there is no codegen to lean on.
 *
 * It is also the one part of the panel that must work when its backend is
 * absent. `isQueryEngineAvailable()` answering `false` is a normal state, not an
 * error.
 */

/** The engines a deployment may run, by the word it is chosen by. */
export const QUERY_ENGINES = ['flow', 'datafusion', 'duckdb'] as const;
export type QueryEngineName = (typeof QUERY_ENGINES)[number];

/** Each engine's name in a sentence somebody reads. */
export const ENGINE_LABEL: Record<QueryEngineName, string> = {
  flow: 'Flow PHP',
  datafusion: 'DataFusion',
  duckdb: 'DuckDB',
};

/**
 * The sentence content written for another engine is shown with, or `null` when
 * it is this deployment's.
 *
 * Shown rather than run (AW-3): text in one engine's language is a syntax error
 * in another's, and the engine's refusal would read as the text being wrong
 * rather than as the deployment being a different one. `deployed` is absent
 * while healthz has not answered, and then nothing is foreign yet.
 */
export function writtenElsewhere(
  written: QueryEngineName,
  deployed: QueryEngineName | null | undefined,
): string | null {
  if (!deployed || written === deployed) return null;
  return `Written for ${ENGINE_LABEL[written]}, and this deployment runs ${ENGINE_LABEL[deployed]}, so it is shown here and not run.`;
}

/**
 * The engine text carried in a URL was written for.
 *
 * A link says, as `writtenFor`; one that does not is Flow's, because that is what
 * every link made before the field held — the rule stored content already lives
 * under. No text at all means the deployed engine's own starter.
 */
export function linkedEngine(
  q: string | undefined,
  writtenFor: QueryEngineName | undefined,
  deployed: QueryEngineName | undefined,
): QueryEngineName {
  if (q === undefined) return deployed ?? 'flow';
  return writtenFor ?? 'flow';
}

const engineSchema = z.object({
  /** Absent from a Flow build that predates AW-3, which is Flow. */
  engine: z.enum(QUERY_ENGINES).optional().default('flow'),
  /** The language a query is written in: `flow-dsl`, `datafusion-python`, `duckdb-python`. */
  language: z.string().optional().default('flow-dsl'),
});

const columnSchema = z.object({ name: z.string(), type: z.string() });

/**
 * A named argument a `read()` may carry, beyond the dataset name.
 *
 * Part of the contract rather than a hint: the aiwatcher API rejects unknown
 * query parameters, so a `read()` argument the dataset never declared turns
 * the whole query into a 400.
 */
const parameterSchema = z.object({
  name: z.string(),
  required: z.boolean(),
  description: z.string(),
  values: z.array(z.string()),
});

const datasetSchema = z.object({
  name: z.string(),
  aliases: z.array(z.string()),
  grain: z.string(),
  description: z.string(),
  requires_run: z.boolean(),
  columns: z.array(columnSchema),
  /** Optional so a panel build stays compatible with an older Flow service. */
  parameters: z.array(parameterSchema).optional().default([]),
});

const datasetsSchema = z.object({
  datasets: z.array(datasetSchema),
  source: z.string(),
  max_rows: z.number(),
});

const resultSchema = z.object({
  columns: z.array(z.string()),
  rows: z.array(z.record(z.string(), z.unknown())),
  row_count: z.number(),
  truncated: z.boolean(),
  /** From `to_output(truncate:)` — whether long cells may be shortened. */
  truncate_cells: z.boolean(),
  dataset: z.string().nullable(),
  grain: z.string().nullable(),
  source: z.string(),
  /** The window the rows were read through, so a short table reads as scoped. */
  window_seconds: z.number().nullable().optional(),
  took_ms: z.number(),
});

const diagnosticSchema = z.object({
  level: z.string(),
  message: z.string(),
  /** Offset into the query as typed — enrichment maps it back for us. */
  offset: z.number(),
  line: z.number(),
  help: z.string().nullable(),
});

const checkSchema = z.object({
  ok: z.boolean(),
  diagnostics: z.array(diagnosticSchema),
  /** Which checkers ran — for Flow, `mago` for syntax and `aiwatcher` for the grammar and schema. */
  checked_by: z.array(z.string()),
});

const errorSchema = z.object({
  error: z.object({
    message: z.string(),
    column: z.number(),
    near: z.string().nullable().optional(),
  }),
});

export type QueryEngineInfo = z.infer<typeof engineSchema>;
export type QueryDiagnostic = z.infer<typeof diagnosticSchema>;
export type QueryCheck = z.infer<typeof checkSchema>;
export type QueryDataset = z.infer<typeof datasetSchema>;
export type QueryParameter = z.infer<typeof parameterSchema>;
export type QueryDatasets = z.infer<typeof datasetsSchema>;
export type QueryResult = z.infer<typeof resultSchema>;

/** A query the engine refused, with where it gave up. */
export class QueryError extends Error {
  constructor(
    message: string,
    readonly column: number,
  ) {
    super(message);
    this.name = 'QueryError';
  }
}

/** Raised when the engine itself is not there. Its own type, because the page treats it differently. */
export class QueryEngineUnavailableError extends Error {
  constructor() {
    super('The query engine is not running.');
    this.name = 'QueryEngineUnavailableError';
  }
}

async function call(path: string, init?: RequestInit): Promise<unknown> {
  let response: Response;

  try {
    response = await fetch(`/query${path}`, init);
  } catch {
    // A refused connection is the engine being absent, not a failed query.
    throw new QueryEngineUnavailableError();
  }

  const body: unknown = await response.json().catch(() => null);
  const serviceError = errorSchema.safeParse(body);
  // A structured upstream failure proves the engine answered. Preserve its
  // reason (e.g. a dataset hub's 502); only an unstructured proxy failure means
  // absent.
  if ([500, 502, 503, 504].includes(response.status) && !serviceError.success) {
    throw new QueryEngineUnavailableError();
  }

  // 404 is a *different* process answering. Every path here is one the engine
  // serves, so it never 404s on its own; what does is whatever else happens to
  // hold the port the proxy points at — measured against a Go service on :8081
  // answering `404 page not found` to everything. Reading that as a refused
  // query blames the pipeline for a port collision, and the message it produces
  // ("The service answered 404") sends the reader to the query they just wrote.
  if (response.status === 404) {
    throw new QueryEngineUnavailableError();
  }

  if (!response.ok) {
    const parsed = serviceError;
    throw new QueryError(
      parsed.success ? parsed.data.error.message : `The service answered ${response.status}.`,
      parsed.success ? parsed.data.error.column : 0,
    );
  }

  return body;
}

export async function isQueryEngineAvailable(): Promise<boolean> {
  try {
    await call('/healthz');
    return true;
  } catch {
    return false;
  }
}

/** Which engine this deployment runs, and the language it reads. */
export async function fetchQueryEngine(): Promise<QueryEngineInfo> {
  return engineSchema.parse(await call('/healthz'));
}

/**
 * The engine this deployment runs, or `null` when there is none.
 *
 * `null` is an answer rather than an error: the engine is optional, and a panel
 * with no engine behind it says so rather than failing.
 */
export function useQueryEngine(): UseQueryResult<QueryEngineInfo | null> {
  return useQuery({
    queryKey: ['query-engine'],
    queryFn: async () => {
      try {
        return await fetchQueryEngine();
      } catch (error) {
        if (error instanceof QueryEngineUnavailableError) return null;
        throw error;
      }
    },
    // Which engine a deployment runs changes with a release, not with a click.
    staleTime: 5 * 60_000,
  });
}

export async function fetchDatasets(): Promise<QueryDatasets> {
  return datasetsSchema.parse(await call('/datasets'));
}

/**
 * What is wrong with a query, without running it.
 *
 * The engine decides what its checkers are. Flow's are two: Mago parses the
 * query as PHP (after the service substitutes the bareword dataset names, which
 * are not valid PHP) and reports where the brackets stopped making sense, and
 * the service's own parser knows the grammar and every column. Neither executes
 * anything.
 */
export async function checkQuery(pipeline: string): Promise<QueryCheck> {
  return checkSchema.parse(
    await call('/check', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ pipeline }),
    }),
  );
}

/**
 * Run a query, scoped to the script's period or the panel's fallback window.
 *
 * `read(default, period: '24h')` is reproducible and wins when present. The
 * request parameter remains a fallback for older/ad-hoc scripts and is
 * forwarded only to aiwatcher routes that accept one.
 */
export async function runQuery(pipeline: string, windowSeconds?: number): Promise<QueryResult> {
  return resultSchema.parse(
    await call('/query', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(
        windowSeconds ? { pipeline, window_seconds: windowSeconds } : { pipeline },
      ),
    }),
  );
}

/**
 * Execute the same plan against a 25-row preview and persist nothing.
 *
 * This has its own endpoint rather than a client-side slice: the cap is
 * applied while the engine reads, so a simulation cannot accidentally walk the
 * full result just to hide most of it in the browser.
 */
export async function simulateQuery(
  pipeline: string,
  windowSeconds?: number,
): Promise<QueryResult> {
  return resultSchema.parse(
    await call('/simulate', {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(
        windowSeconds ? { pipeline, window_seconds: windowSeconds } : { pipeline },
      ),
    }),
  );
}
